habitat_code/ramachandran_code_for_review.py

```diff
@@ -0,0 +1,511 @@
+
+boundary_url = ("https://gis.bouldercolorado.gov/ags_svr1/rest/services/plan/CityLimits/MapServer/0/query?outFields=*&where=1%3D1&f=geojson")
+
+boundary_path = os.path.join(data_dir, 'city_boundary.geojson')
+
+if os.path.exists(boundary_path):
+
+    boundary_gdf = gpd.read_file(boundary_path)
+    print("Data is already downloaded.")
+
+else:
+    # Mimic web browser
+    user_agent = (
+        'Mozilla/5.0 (X11; Linux x86_64; rv:60.0) '
+        'Gecko/20100101 Firefox/81.0'
+    )
+
+    # Download GEOJSON
+    r = requests.get(url=boundary_url, headers={'User-Agent': user_agent})
+
+    # Read GeoJSON data into a GeoDataFrame
+    boundary_geojson_data = r.json()
+
+    # Cache
+    boundary_gdf = (gpd.GeoDataFrame
+                    .from_features(boundary_geojson_data['features'])
+                    .to_file(boundary_path, driver="GeoJSON")
+                    )
+    
+    # Read
+    boundary_gdf = gpd.read_file(boundary_path)
+    print("Data downloaded and loaded.")
+
+
+
+# %%
+boundary_gdf
+
+# %%
+boundary_gdf.to_crs("4326").plot()
+
+# %%
+# Download census tracts
+print(us.states.CO.fips)
+
+colorado_tracts_url = ("https://www2.census.gov/geo/tiger/TIGER2023/"
+           "TRACT/tl_2023_08_tract.zip")
+
+colorado_tracts_path = os.path.join(data_dir, 'colorado_census_tracts.geojson')
+
+if os.path.exists(colorado_tracts_path):
+    colorado_tracts = gpd.read_file(colorado_tracts_path).to_crs("4326")
+    print("Data is already downloaded.")
+else:
+    gpd.read_file(colorado_tracts_url).to_file(colorado_tracts_path, driver="GeoJSON")
+    colorado_tracts_gdf = gpd.read_file(colorado_tracts_path).to_crs("4326")
+
+# %%
+colorado_tracts.plot(edgecolor="white", linewidth=0.5)
+
+# %%
+# Select only the census tracts in Boulder
+print(colorado_tracts.crs)
+print(boundary_gdf.crs)
+
+boulder_tracts_gdf = gpd.sjoin(colorado_tracts, boundary_gdf, how="inner", predicate="intersects")
+boulder_tracts_gdf.plot(edgecolor="white", linewidth=0.5)
+
+# Clip to city boundary 
+boulder_city_tracts_gdf = gpd.clip(boulder_tracts_gdf, boundary_gdf)
+boulder_city_tracts_gdf.plot(edgecolor="white", linewidth=0.5)
+
+# %%
+# Download census data
+
+# B06011_001E: Estimate!!Median income in the past 12 months --!!Total:	
+# Sources: https://api.census.gov/data/2019/acs/acs5/variables.html; https://pypi.org/project/census/
+
+# Obtain Census variables from the 2019 American Community Survey at the tract level for Illinois (FIPS code: 17)
+census_path = os.path.join(data_dir, 'census_data_for_colorado.csv')
+
+if os.path.exists(census_path):
+    census_df = pd.read_csv(census_path)
+    print("Data is already downloaded.")
+
+else:
+    # Authenticate
+    api_key = getpass.getpass('U.S. Census API Key')
+    c = Census(api_key)
+    c
+
+    CO_census = c.acs5.state_county_tract(fields = ('NAME', 'B06011_001E'),
+                                      state_fips = '08',
+                                      county_fips = "*",
```

Comment:
You can supply the County FIPS code here instead of '*' to reduce the data handled. The  us python package lets you do this with the following code: 
```python
county_fips = us.states.lookup(state).fips + us.counties.lookup(county_name).fips
```

Reply:
🔥